examples/tests/test_01_provider_fastapi.py

```diff
@@ -0,0 +1,142 @@
+Calls to this endpoint mock the relevant database calls to set the provider into
+the correct state.
+"""
+
+from __future__ import annotations
+
+from multiprocessing import Process
+from typing import Any, Generator
+from unittest.mock import MagicMock
+
+import pytest
+import uvicorn
+from examples.src.fastapi import app
+from pact import Verifier
+from pydantic import BaseModel
+from yarl import URL
+
+PROVIDER_URL = URL("http://localhost:8080")
+
+
+class ProviderState(BaseModel):
+    """Define the provider state."""
+
+    consumer: str
+    state: str
+
+
+@app.post("/_pact/provider_states")
+async def mock_pact_provider_states(state: ProviderState) -> dict[str, str | None]:
+    """
+    Define the provider state.
+
+    For Pact to be able to correctly tests compliance with the contract, the
+    internal state of the provider needs to be set up correctly. Naively, this
+    would be achieved by setting up the database with the correct data for the
+    test, but this can be slow and error-prone. Instead this is best achieved by
+    mocking the relevant calls to the database so as to avoid any side effects.
+
+    For Pact to be able to correctly get the provider into the correct state,
+    this function is used to define an additional endpoint on the provider. This
+    endpoint is called by Pact before each test to ensure that the provider is
+    in the correct state.
+    """
+    mapping = {
+        "user 123 doesn't exist": mock_user_123_doesnt_exist,
+        "user 123 exists": mock_user_123_exists,
+    }
+    return {"result": mapping[state.state]()}
+
+
+def run_server() -> None:
+    """
+    Run the FastAPI server.
+
+    This function is required to run the FastAPI server in a separate process. A
+    lambda cannot be used as the target of a `multiprocessing.Process` as it
+    cannot be pickled.
+    """
+    uvicorn.run(app, host=PROVIDER_URL.host, port=PROVIDER_URL.port)
+
+
+@pytest.fixture(scope="module")
+def verifier() -> Generator[Verifier, Any, None]:
+    """Set up the Pact verifier."""
+    proc = Process(target=run_server, daemon=True)
+    verifier = Verifier(
+        provider="UserProvider",
+        provider_base_url=str(PROVIDER_URL),
+    )
+    proc.start()
+    yield verifier
+    proc.kill()
+
+
+def mock_user_123_doesnt_exist() -> None:
+    """Mock the database for the user 123 doesn't exist state."""
+    import examples.src.fastapi
+
+    examples.src.fastapi.FAKE_DB = MagicMock()
+    examples.src.fastapi.FAKE_DB.get.return_value = None
+
+
+def mock_user_123_exists() -> None:
+    """
+    Mock the database for the user 123 exists state.
+
+    You may notice that the return value here differs from the consumer's
+    expected response. This is because the consumer's expected response is
+    guided by what the consumer users.
+
+    By using consumer-driven contracts and testing the provider against the
+    consumer's contract, we can ensure that the provider is only providing what
```

Comment:
sentence needs finishing?

we can ensure that the provider is able to safely evolve their api, whilst ensuring them meet each consumers expectations.

- Also this is a neat point for mentioning and I referred to in an earlier comment on the consumer side

Reply:
Good catch, my bad!